samples/src/Aspects.Lazy/readme.md

```diff
@@ -0,0 +1,11 @@
+﻿This package provides Lazy proerty.
+Put `LazyAttribute` attribure on your get only properties. 
+
+```csharp
+class TestClass
+{
+    [Lazy]
+    public ServiceA ServiceA => new ServiceA(DateTime.Now);

```

Comment:
But this isn't Lazy in fact. I mean the ```new ServiceA(DateTime.Now)``` is executed when ```TestClass``` constructor is called, not when you access the property ```ServiceA``` for the first time. so it works as rather as a cache. Correct me if I've mistaken!

Reply:
Yes, you are right.

I need your help. Can you implement it like this:

```csharp
public class Test
{
    [Lazy]
    public string StrService => $"{DateTime.Now:O}";
}
```

```csharp
private Lazy<string> <StrService>_Lazy_Fody_BackingField;

public string StrService => <StrService>_Lazy_Fody_BackingField.Value;

public string <StrService>_Lazy_Fody_Method()
{
 return $"{DateTime.Now:O}";
}

public Test()
{
 <StrService>_Lazy_Fody_BackingField = new Lazy<string>(() => $"{DateTime.Now:O}", LazyThreadSafetyMode.None);
}
```

Reply:
Unfortunately the core doesn't have the feature for preventing property to be initialized. Because backfield is set after the value is calculated, that's why overriding getter doesn't really work.

I'm afraid this is kind of feature requires heavy core changes, so it is able to override particular instructions in the method code. 
At this time we have no plans to implement it. But any contribution are welcome!
Though at the moment for this feature I'd look at Fody ))

Reply:
The Fody add-in: https://github.com/tom-englert/Lazy.Fody

I'm just getting started with IL wave, learn from you.

Reply:
Fody is good project too. And even though AspectInjector and Fody have some features in common, the idea behind AspectInjector is to allow users write aspects without having to deal with IL code.
So that's why only AspectInjector core has IL manipulations.  The result is however that some features is impossible to implement with AspectInjector, for example the ```Lazy``` one that you've suggested.
You could try to figure how make Property initialization manipulation generic enough so we can add this implementation in the core. For example we could introduce new advice target ```Target.PropertyInit``` on which the core modifies target class constructor code so user can do with this property initialization whatever he wants.

Reply:
I have a question.

> I mean the new ServiceA(DateTime.Now) is executed when TestClass constructor is called, not when you access the property ServiceA for the first time. 

I think, get only property is different and auto property, get only property is not init in constructor.

```csharp
public string StrService => $"{DateTime.Now:O}";
```

There is IL code:

```
.method public hidebysig specialname rtspecialname 
instance void .ctor () cil managed 
{
    // Header Size: 1 byte
    // Code Size: 8 (0x8) bytes
    .maxstack 8
    
    /* 0x000002AC */ IL_0000: ldarg.0
    /* 0x000002AD */ IL_0001: call      instance void [System.Runtime]System.Object::.ctor()
    /* 0x000002B2 */ IL_0006: nop
    /* 0x000002B3 */ IL_0007: ret
}

.property instance string ServiceC()
{
    // Token: 0x06000036 RID: 54 RVA: 0x00002A38 File Offset: 0x00000C38
    .get instance string AspectInjectorTest.TestClass::get_ServiceC()
}
```

I think it's just like a method, there is no back filed.
Am I right?

Reply:
omg, you're actually right, I double-checked. So i fact you can make Lazy-like get-only-properties! I will only work for _arrow-properties_ but still

<details>
  <summary>My test</summary>

```c#
using AspectInjector.Broker;
using System;
using System.Collections.Concurrent;
using System.Reflection;

namespace test_getonlyproperty
{
    class Program
    {
        static void Main(string[] args)
        {
            for (int i = 0; i < 2; i++)
            {
                Console.WriteLine("$$Test " + i);

                Console.WriteLine("## Instancing class");
                var a = new TestClass();
                Console.WriteLine("## First run");
                a.Do();
                Console.WriteLine("## Second run");
                a.Do();

                Console.WriteLine();
                Console.WriteLine();
            }            
        }
    }

    [Aspect(Scope.PerInstance)]
    [Injection(typeof(LazyAttribute))]
    [AttributeUsage(AttributeTargets.Property, AllowMultiple = false)]
    public class LazyAttribute : Attribute
    {
        private readonly ConcurrentDictionary<MethodBase, object> _instances = new ConcurrentDictionary<MethodBase, object>();

        [Advice(Kind.Around)]
        public object Around(
            [Argument(Source.Arguments)] object[] args,
            [Argument(Source.Target)] Func<object[], object> target,
            [Argument(Source.Metadata)] MethodBase metadata)
        {
            var result = _instances.GetOrAdd(metadata, m => target(args));
            return result;
        }
    }

    class TestClass
    {
        public string Prop1 => GetInit(nameof(Prop1));
        [Lazy]
        public string Prop2 => GetInit(nameof(Prop2));
        public string Prop3 { get; } = GetInit(nameof(Prop3));


        public static string Prop4 => GetInit(nameof(Prop4));

        [Lazy]
        public static string Prop5 => GetInit(nameof(Prop5));
        public static string Prop6 { get; } = GetInit(nameof(Prop6));

        public TestClass()
        {
            Console.WriteLine(".ctor");
        }

        static TestClass()
        {
            Console.WriteLine(".cctor");
        }

        internal void Do()
        {
            Console.WriteLine("method");

            var a = Prop1 + Prop2 + Prop3 + Prop4 + Prop5 + Prop6;

            Console.WriteLine("result: " + a);
        }

        private static string GetInit(string v)
        {
            Console.WriteLine("init: " + v);
            return v;
        }
    }
}
  ```
  
</details>

Reply:
So this will definitely fly! If you could address these points:

- bakfields are better done (I think, but may be wrong) by MethodBase as a key, as it holds info about type, and method, and generics
- think about making sure real property called only once -> you may think of ```lock()``` or other things
- documentation that states that only arrow-properties will be real lazy, other properties will be just cached but not lazy